### ISSUE TITLE
Recording received videotracks

### ISSUE THREAD

**Issue by ffittschen** [twilio/video-quickstart-ios#123]
Hi,

we are currently trying to record received video tracks (`TVIVideoTrack`), but there seems to be something off with the pixel format.

For testing purposes we used the [MovieRecorder.m](https://developer.apple.com/library/content/samplecode/VideoSnake/Listings/Classes_MovieRecorder_m.html#//apple_ref/doc/uid/DTS40012327-Classes_MovieRecorder_m-DontLinkElementID_6) of Apples VideoSnake demo and created a custom `RecordingRenderer`, that creates a `CMSampleBuffer` of the `imageBuffer` of the `frame` in `renderFrame(_:)` and passes it to the `MovieRecorder`. The result of appending the sampleBuffer to the AVAssetWriterInput is always false:
```
let success = input.appendSampleBuffer(sampleBuffer)
if !success {
    print(assetWriter.error)
}
```
And the error property of the AVAssetWriter doesn't provide helpful error messages:
> AVAssetWriterError: Error Domain=AVFoundationErrorDomain Code=-11800 "The operation could not be completed" UserInfo={NSUnderlyingError=0x13ad3c470 {Error Domain=NSOSStatusErrorDomain Code=-12780 "(null)"}, NSLocalizedFailureReason=An unknown error occurred (-12780), NSLocalizedDescription=The operation could not be completed}

However:
When I instantiated a "fake" `TVIVideoFrame` manually by creating a `CVPixelBuffer` from a `UIImage` and calling `TVIVideoFrame(timestamp: 0, buffer: fakePixelBuffer, orientation: TVIVideoOrientation.up)`, the video will be recorded without problems!

Is the `imageBuffer` of the frame always in `TVIPixelFormatYUV420PlanarFullRange` format if I don't specify any `optionalPixelFormats` in the `TVIVideoRenderer`?

**Comment by ceaglest** [comment]
Hi @ffittschen,

Thanks for asking about recording using AVAssetWriter. We are a little bit light on sample code in this area, so I'm happy to help out as best I can.

> When I instantiated a "fake" TVIVideoFrame manually by creating a CVPixelBuffer from a UIImage and calling TVIVideoFrame(timestamp: 0, buffer: fakePixelBuffer, orientation: TVIVideoOrientation.up), the video will be recorded without problems!

This makes sense, AVAssetWriter should be well equipped to handle Apple's preferred pixel formats like BGRA, ARGB, and NV12.

> Is the imageBuffer of the frame always in TVIPixelFormatYUV420PlanarFullRange format if I don't specify any optionalPixelFormats in the TVIVideoRenderer?

Yes.

`TVIPixelFormatYUV420PlanarFullRange` or I420 is the working format of WebRTC's media engine. This is why its a required format for renderers - we can always convert from any `TVIPixelFormat` **to** I420. Unfortunately we don't support arbitrary format conversions **from** I420 to your pixel format of choice. For remote video you should always expect frames to be delivered in I420, because this is the format that the VP8 decoder uses. This will change in the future once we support video codecs like [H.264](https://github.com/twilio/video-quickstart-swift/issues/88), which decode to NV12 natively on iOS. We are also considering decoding VP8 to NV12 in the future, because this format is more compatible with the iOS / Apple ecosystem.

For local video it really depends on what format your `TVIVideoCapturer` provides, for example `TVICameraCapturer` will capture in NV12, and `TVIScreenCapturer` in ARGB. If your renderer supports these formats you can receive frames without conversion. If not, you will get I420.

If you want to record remote video using AVAssetWriter you may need to perform the format conversion from I420 to NV12 yourself. This is as "simple" as interleaving the U and V planes into a combined UV plane, and copying the Y plane without modification (less padding differences). I'm sorry that we don't have sample code that demonstrates this conversion. In the meantime, I can share some code I wrote a few years ago for [PerchRTC](https://github.com/perchco/perchrtc/blob/master/PerchRTC/Renderers/PHFrameConverter.m#L566). This method attempts to either copy NV12 to NV12, or repack I420 to NV12 and might give you some ideas to get started.

Best,
Chris Eagleston

**Comment by ffittschen** [comment]
Thanks for the quick reply @ceaglest! 
Okay, I'll try to convert from I420 to a format that works better with the `AVAssetWriter`. When I managed to get it working, I'll post the methods here or link to a gist in case someone else has the same problems

**Comment by ceaglest** [comment]
No problem @ffittschen. If you get something working feel free to share the snippets. At some point we would like to offer sample code showing local recording, and could use yours as a jumping off point. As a quick sanity check, you might want to try recording local video from TVICameraCapturer (opting into NV12 on your renderer).

Also, we are working towards releasing our Group Rooms beta in the next few weeks. This product is currently in limited developer preview, and will offer server side recording of audio/video streams.

**Comment by ffittschen** [comment]
It took a while, but I managed to get it working!
**TL;DR** Converting the frames from `I420` to `NV12` allows me to append them to an `AVAssetWriterInput`.

The steps I took were the following: Compile the current version of the [libyuv](https://chromium.googlesource.com/libyuv/libyuv/) project into a framework. To do so I forked caojianhua/libyuv-iOS to [ffittschen/libyuv-iOS](https://github.com/ffittschen/libyuv-iOS) and replaced the contents of the libyuv folder with the current version.

After embedding the `libyuv.framework` into my main Xcode Project, I can import it and call the `I420ToNV12` function:
```swift
import TwilioVideo
import libyuv

class RecordingRenderer: NSObject {
    // ...
    @discardableResult
    fileprivate func convertI420toNV12(fromFrame frame: TVIVideoFrame, toPixelBuffer pixelBuffer: CVPixelBuffer) -> Bool {
        let sourceFormat = frame.imageBuffer.pixelFormat
        let destinationFormat = CVPixelBufferGetPixelFormatType(pixelBuffer)
        
        guard destinationFormat == kCVPixelFormatType_420YpCbCr8BiPlanarFullRange else {
            preconditionFailure("Destination format is not bi-planar! Check the attributes of the pixelBufferPool.")
        }
        guard sourceFormat == kCVPixelFormatType_420YpCbCr8Planar || sourceFormat == kCVPixelFormatType_420YpCbCr8PlanarFullRange else {
            preconditionFailure("Source is not planar! Can't pack or copy these pixels!")
        }
        
        CVPixelBufferLockBaseAddress(frame.imageBuffer, CVPixelBufferLockFlags(rawValue: 0))
        CVPixelBufferLockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
        
        defer {
            CVPixelBufferUnlockBaseAddress(frame.imageBuffer, CVPixelBufferLockFlags(rawValue: 0))
            CVPixelBufferUnlockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
        }
        
        // swiftlint:disable operator_usage_whitespace line_length
        let src_yAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 0)?.assumingMemoryBound(to: UInt8.self)
        let src_uAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 1)?.assumingMemoryBound(to: UInt8.self)
        let src_vAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 2)?.assumingMemoryBound(to: UInt8.self)
        let dst_yAddress  = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 0)?.assumingMemoryBound(to: UInt8.self)
        let dst_uvAddress = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 1)?.assumingMemoryBound(to: UInt8.self)
        let src_yBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 0))
        let src_uBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 1))
        let src_vBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 2))
        let dst_yBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0))
        let dst_uvBytesPerRow = Int32(CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 1))
        let width = Int32(frame.width)
        let height = Int32(frame.height)
        // swiftlint:enable operator_usage_whitespace line_length
        
        let success = I420ToNV12(src_yAddress,
                                 src_yBytesPerRow,
                                 src_uAddress,
                                 src_uBytesPerRow,
                                 src_vAddress,
                                 src_vBytesPerRow,
                                 dst_yAddress,
                                 dst_yBytesPerRow,
                                 dst_uvAddress,
                                 dst_uvBytesPerRow,
                                 width,
                                 height
        )
        
        return success == 0
    }
    // ...
}
extension RecordingRenderer: TVIVideoRenderer {
    func renderFrame(_ frame: TVIVideoFrame) {
        guard let pixelBuffer = dequeuePixelBuffer(forFrame: frame) else {
            log.warning("Dropping frame! BufferPool is maxed out")
            return
        }
        
        convertI420toNV12(fromFrame: frame, toPixelBuffer: pixelBuffer)
        
        let presentationTime = CMTime(value: frame.timestamp, timescale: 1000000)
        
        streamRecorder.appendVideo(pixelBuffer, with: presentationTime)
    }
}
```

**Comment by ceaglest** [comment]
Very cool, thanks for sharing @ffittschen!

WebRTC uses libyuv internally so we have access to these functions within our SDK. Unfortunately it isn't something that we expose in our public APIs, but your instinct to use libyuv was probably the right one.

We would still like to publish sample code demonstrating recording, but until then I hope others can learn from the work you shared.

Best,
Chris

**Comment by MBuLoft1** [comment]
Thank you for the sample code ! We did a variant that goes to a BGRA image, using libyuv, so in case this may help someone, here is the difference in the call (assuming a pre-allocated pixelBuffer (class CVPixelBuffer) to size width*4 (bytes) :

```
            let srcYAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 0)?.assumingMemoryBound(to: UInt8.self)
            let srcUAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 1)?.assumingMemoryBound(to: UInt8.self)
            let srcVAddress  = CVPixelBufferGetBaseAddressOfPlane(frame.imageBuffer, 2)?.assumingMemoryBound(to: UInt8.self)
            let srcYBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 0))
            let srcUBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 1))
            let srcVBytesPerRow  = Int32(CVPixelBufferGetBytesPerRowOfPlane(frame.imageBuffer, 2))
            let destAddress = CVPixelBufferGetBaseAddress(pixelBuffer!)
            let width = Int32(frame.width)
            let height = Int32(frame.height)
            let result = I420ToBGRA(srcYAddress, srcYBytesPerRow,
                                    srcUAddress, srcUBytesPerRow,
                                    srcVAddress, srcVBytesPerRow,
                                    destAddress, Int32(frame.width*4),
                                    width, height);

            // From here can use the Apple APIs to work with images
            let ciImage = CIImage(cvPixelBuffer: pixelBuffer!, options: nil)
            let ciContext = CIContext(options: nil)
            let size =  CVImageBufferGetCleanRect(imageBuffer)
            let cgImage: CGImage = ciContext.createCGImage(ciImage, from: size)!
            // your image processing code here...
```
In our case integrating the libyuv in our codebase was straightforward because we already maintain a C++ image engine, so we added libyuv to that. Otherwise starting from libyuv-IOS as was mentioned above should work as well.

**Comment by piyushtank** [comment]
Closing an old ticket. Feel free to reopen or file another ticket if you run into any issues.
